gems/aws-sdk-core/lib/aws-sdk-core/shared_credentials.rb

```diff
@@ -25,57 +26,82 @@ class SharedCredentials
+    #
+    # @option [Boolean] :enable_refresh If true, the credentials will be automatically
+    #   refreshed when they are near expiration. Defaults to false.
+    #
     def initialize(options = {})
-      shared_config = Aws.shared_config
-      @path = options[:path]
-      @path ||= shared_config.credentials_path
-      @profile_name = options[:profile_name]
-      @profile_name ||= ENV['AWS_PROFILE']
-      @profile_name ||= shared_config.profile_name
-      if @path && @path == shared_config.credentials_path
-        @credentials = shared_config.credentials(profile: @profile_name)
-      else
-        config = SharedConfig.new(
-          credentials_path: @path,
-          profile_name: @profile_name
-        )
-        @credentials = config.credentials(profile: @profile_name)
-      end
+      @enable_refresh = options[:enable_refresh] || false # Refresh disabled by default
+      @refresh_interval = options[:refresh_interval] || 10 * 60 # Default refresh interval: 10 minutes
+      @path = options[:path] || Aws.shared_config.credentials_path
+      @profile_name = options[:profile_name] || ENV['AWS_PROFILE'] || Aws.shared_config.profile_name
+
+      super(options)
     end
 
-    # @return [String]
+    # @return [String] The path to the credentials file
     attr_reader :path
 
-    # @return [String]
+    # @return [String] The name of the profile being used
     attr_reader :profile_name
 
-    # @return [Credentials]
-    attr_reader :credentials
+    # @return [Credentials] The loaded credentials
+    def credentials
+      refresh_if_necessary
+      @credentials
+    end
 
     # @api private
     def inspect
       parts = [
         self.class.name,
         "profile_name=#{profile_name.inspect}",
-        "path=#{path.inspect}",
+        "path=#{path.inspect}"
       ]
       "#<#{parts.join(' ')}>"
     end
 
-    # @deprecated This method is no longer used.
-    # @return [Boolean] Returns `true` if a credential file
-    #   exists and has appropriate read permissions at {#path}.
-    # @note This method does not indicate if the file found at {#path}
-    #   will be parsable, only if it can be read.
-    def loadable?
-      !path.nil? && File.exist?(path) && File.readable?(path)
+    # Refreshes credentials from the shared credentials file.
+    def refresh
+      shared_config = Aws.shared_config
+
+      if @path && @path == shared_config.credentials_path
+        # Use credentials from shared config if paths match
+        @credentials = shared_config.credentials(profile: @profile_name)
+      else
+        # Load credentials from specified path and profile
+        config = SharedConfig.new(
+          credentials_path: @path,
+          profile_name: @profile_name
+        )
+        @credentials = config.credentials(profile: @profile_name)
+      end
+
+      # Set expiration time if credentials are present
+      @expiration = @credentials ? (Time.now + @refresh_interval) : nil
```

Comment:
We may not actually want to set this as it has implications for *using* the credentials (Eg: the credential's expiration is used when creating a presigned url).  Instead we may need an internal variable to track this.

Reply:
My original comment may have been misleading. I think this is right.